Reintroduce optional parameter versioning mechanism for airframe maintainers

MaEtUgR opened the pull request:
### Solved Problem
When working on multiple projects I found that many reintroduced a mechanism to allow automatically resetting the parameter configuration during a product update. This functionality was initially introduced in https://github.com/PX4/PX4-Autopilot/pull/11760 and removed again in https://github.com/PX4/PX4-Autopilot/pull/17279 because we (including me) thought that the `param set-default` mechanism of https://github.com/PX4/PX4-Autopilot/pull/16796 would solve all the cases. But the case where the airframe maintainer wants to enforce a reset to airframe defaults is not covered anymore with the `param set-default` mechanism. For products based on PX4 this is still required to ensure proper functionality after a major update.

### Solution
I suggest to reintroduce what was removed in https://github.com/PX4/PX4-Autopilot/pull/17279 in a slightly adapted way.
The mechanism is still exactly the same. An airframe maintainer can **optionally** add e.g. an `set PARAM_DEFAULTS_VER 2` line in the airframe file such that vehicles coming from an older (or newer) version airframe as detected using the `SYS_PARAM_VER` parameter get reset to the default configuration. As usual accumulating flight numbers, hours, RC, sensor and **newly also airspeed calibration** are preserved.

### Changelog Entry
```
Feature: Reintroduce optional parameter versioning mechanism for airframe maintainers
```

### Alternatives
I tried to work around the necessity to trigger a reboot for the mechanism but after some testing it's still the most useful way in my eyes. The parameter version is only available after loading the airframe but arch and board defaults need to be loaded before that and the reset has to be done before that even because otherwise all `param set` lines in these configurations would be reset again. So either we'd load the airframe twice per boot or just reboot the system if a reset is required.

### Test coverage
I tested this on a vehicle which already had a parameter version from the previous mechanism and it seamlessly reset to the latest defaults even if I manually changed the parameters manually before the update.

### Context
I was in discussion with @dagar about a possibly better less shotgun way to address the specific use cases solved by this mechanism but at least to me it's not clear yet how to solve all cases especially during bringups and major updates where a known working configuration is usually crucial to move multiple vehicles forward.

dagar commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -175,13 +180,11 @@ else
-		# Reset params except Airframe, RC calibration, sensor calibration, flight modes, total flight time, and next flight UUID.
-		param reset_all SYS_AUTOSTART RC* CAL_* COM_FLTMODE* LND_FLIGHT* TC_* COM_FLIGHT*
+		# Reset parameters except airframe, parameter version, RC calibration + flight modes, sensor + airspeed calibration, total flight time, flight UUID
+		param reset_all SYS_AUTOS* SYS_PA* RC* COM_FLTMODE* CAL_* ASPD_S* LND_FLIGHT* TC_* COM_FLIGHT*
```
ASPD_SCALE_APPLY?

sfuhrer replied:
@MaEtUgR I guess we could do
```
param reset_all ASPD_S*
param reset ASPD_SCALE_APPLY ASPD_SCALE_NSD
```
Ugly but until we have better separation between configuration, settings and auto-adjusted settings in the params a way forward at least.

MaEtUgR replied:
I removed this change from the pr since the goal is merely about the param versioning and not changing the entire reset logic. I should have done that from the beginning then.

dagar commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -175,13 +180,11 @@ else
-		# Reset params except Airframe, RC calibration, sensor calibration, flight modes, total flight time, and next flight UUID.
-		param reset_all SYS_AUTOSTART RC* CAL_* COM_FLTMODE* LND_FLIGHT* TC_* COM_FLIGHT*
+		# Reset parameters except airframe, parameter version, RC calibration + flight modes, sensor + airspeed calibration, total flight time, flight UUID
+		param reset_all SYS_AUTOS* SYS_PA* RC* COM_FLTMODE* CAL_* ASPD_S* LND_FLIGHT* TC_* COM_FLIGHT*
```
```suggestion
		param reset_all SYS_AUTOS* SYS_PA* RC* COM_FLTMODE* CAL_* ASPD_S* LND_FLIGHT* TC_* COM_FLIGHT_UUID
```

MaEtUgR replied:
`COM_FLI*` saves even more characters! 😄

MaEtUgR replied:
Removed the shortenings. The reason I even considered them is because we had it before that this line got too long to be executed with some additional customizations. Of course there are better solutions like a proper categorization. Shortening with wildcards is just a low hanging fruit.

sfuhrer commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -175,13 +180,11 @@ else
-		# Reset params except Airframe, RC calibration, sensor calibration, flight modes, total flight time, and next flight UUID.
-		param reset_all SYS_AUTOSTART RC* CAL_* COM_FLTMODE* LND_FLIGHT* TC_* COM_FLIGHT*
+		# Reset parameters except airframe, parameter version, RC calibration + flight modes, sensor + airspeed calibration, total flight time, flight UUID
+		param reset_all SYS_AUTOS* SYS_PA* RC* COM_FLTMODE* CAL_* ASPD_S* LND_FLIGHT* TC_* COM_FLIGHT*
```
What's actually the argument for not resetting COM_FLTMODE?

MaEtUgR replied:
I would vote for resetting it. AFAIK it was purely left for convenience.

MaEtUgR replied:
Convenience but I'm not chaing it in this pr since it's not the goal here.

dagar commented:
We could add something to make it easy to skip the `volatile` parameters.

dagar commented:
How about `param reset_all --skip-volatile X`? Then we'd only need TC and a few other little things on top.

MaEtUgR commented:
I
- rebased on main
- removed the unrelated cahnges like shortening the reset exclusion parameter name wildcards and adding the airspeed calibration to the exclusions

This pr is about the parameter versioning mechanism and not about airspeed calibration or revising the volatile parameter reset mechanism.

sfuhrer commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -36,6 +36,11 @@ set SDCARD_FORMAT no
+# Airframe parameter versioning: airframe maintainers can optionally set this in the
+# airframe startup script, and then increase it by one whenever they want to ensure
+# that parameters are reset to the airframe defaults for an update.
+set PARAM_DEFAULTS_VER 1
```
```suggestion
# Set to 1 by default, can get overwritten in the airframe startup script (config file).
set PARAM_DEFAULTS_VER 1
```

MaEtUgR replied:
I wrote it slightly differently to not contain redundant statements but the points you suggested to add. See d7437b4

sfuhrer commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -226,6 +229,17 @@ else
 	fi
 	unset AUTOSTART_PATH
 
+	# Check parameter version and reset upon airframe configuration version mismatch
```
```suggestion
	# Check parameter version and reset upon airframe configuration version mismatch.
```

MaEtUgR replied:
Fixed in d7437b4

sfuhrer commented on ROMFS/px4fmu_common/init.d/rcS:
```diff
@@ -226,6 +229,17 @@ else
 	unset AUTOSTART_PATH
 
+	# Check parameter version and reset upon airframe configuration version mismatch
+	# reboot required because "param reset_all" would reset all "param set" lines from airframe
```
```suggestion
	# Reboot required because "param reset_all" would reset all "param set" lines from airframe.
```

MaEtUgR replied:
Fixed in d7437b4

sfuhrer commented on src/lib/systemlib/system_params.c:
```diff
@@ -83,6 +83,19 @@ PARAM_DEFINE_INT32(SYS_AUTOCONFIG, 0);
+ *
+ * This is used internally only: an airframe configuration might set an expected
+ * parameter version value via PARAM_DEFAULTS_VER. This is checked on bootup
+ * against SYS_PARAM_VER, and if they do not match, parameters from the airframe
```
```suggestion
 * against SYS_PARAM_VER, and if they do not match, parameters are reset and
```

sfuhrer commented on src/lib/systemlib/system_params.c:
```diff
@@ -83,6 +83,19 @@ PARAM_DEFINE_INT32(SYS_AUTOCONFIG, 0);
+ * This is used internally only: an airframe configuration might set an expected
+ * parameter version value via PARAM_DEFAULTS_VER. This is checked on bootup
+ * against SYS_PARAM_VER, and if they do not match, parameters from the airframe
+ * configuration are reloaded.
```
```suggestion
 * the ones in the airframe configuration are reloaded.
```

sfuhrer reviewed (dismissed):
Thanks for coming back to this - I agree that we have to focus on bringing the reset mechanism in here, and should move the work around which params to not reset to another PR.
I just have some proposals for cleaner comments. Otherwise looks good to me (haven't tried it out myself though).

hamishwillee commented:
FYI There have been some questions in the last couple of days on discord about how parameters "change managed" across versions, following renames and deletions. Would appreciate an update to the param docs for that.

MaEtUgR commented:
@sfuhrer Thanks for your suggestions, it's important to be clear 🙏 I tried to adjust the wording from what we originally had to address the valid input.

MaEtUgR commented:
Rebased on main to have CI run again. The previous errors looked like the ones we fixed recently, see #23200

sfuhrer commented:
69 bytes over flash on v5
